include/fiction/algorithms/simulation/sidb/ground_state_space.hpp

```diff
@@ -0,0 +1,833 @@
+ * fixpoint" concepts to prune the simulation search space for not only a flat layout of SiDBs, but rather generalizes,
+ * and lifts the physically informed space pruning technique introduced with *QuickExact* to higher order, allowing
+ * *Ground State Space* to prune multiset charge state configurations at any level in a cluster hierarchy.
+ *
+ * The role of the cluster hierarchy is to rank interactions between groups, or clusters of SiDBs that together make up
+ * the whole layout, such that the variation in electrostatic potential under different charge state assignments is
+ * highest between the children clusters of clusters that low in the hierarchy. Thereby, the structure allows us to
+ * consider the most charge state assignment-dependent interaction in a more detailed physically informed space pruning
+ * analysis, enabling high pruning efficacy for the few pruning tests (with respect to the exponential search space).
+ *
+ * Starting at a clustering of all singleton clusters, the charge spaces, ie. a set of multiset charge configurations
+ * (initially { {{-}}, {{0}}, {{+}} } or omitting the singleton multiset {{+}} in the case of base 2 pre-simulation),
+ * are pruned iteratively through potential bound analysis. Through merges, ie., replacing a set of children in the
+ * clustering with their parent, we may inspect the most crucially dependant interactions in the layout separately. The
+ * procedure finishes when the charge spaces have been folded all the way up to the top cluster, parent of all, which
+ * then contains all information resulting from the construction. *ClusterComplete*, without much trickery, now simply
+ * unfolds this result, allowing simulation of problems that were previously seen as astronomical, due to the
+ * (base 2 or 3) exponential growth in the number of SiDBs.
+ *
+ * @tparam Lyt SiDB cell-level layout type.
+ * @param lyt Layout to construct the ground state space of.
+ * @param max_cluster_size_for_witness_partitioning This specifies the maximum cluster size for which *Ground State
+ * Space* will solve an NP-complete sub-problem exhaustively. The sets of SiDBs that witness local population stability
+ * for each respective charge state may be partitioned into disjoint sets such that the number of required witnesses for
+ * each respective charge state is satisfied. If no such partition exists, the multiset charge configuration associated
+ * with the requirements may be rejected. Defaulted to `6` to avoid impairing runtimes, while allowing some pruning.
+ * @param phys_params The physical parameters that *Ground State Space* will use to prune simulation search space.
+ * @return The results of the construction, which include the top cluster which parents all other clusters, and thereby
+ * contains the charge spaces of each cluster.
+ */
+template <typename Lyt>
+ground_state_space_stats
+ground_state_space(const Lyt& lyt, const uint64_t max_cluster_size_for_witness_partitioning = 6,
```

Comment:
can you explain/mention why ``6`` is used as default for ``max_cluster_size_for_witness_partitioning``?

Reply:
I added another parameter for GSS too earlier. I'll write a little bit about their default settings which will convey the following message:

The default values are chosen to not to impair runtimes at all (as found empirically), and these parameters should be taken into consideration for very large simulations mostly. In those cases, the user may use the GSS `report` functionality along with the two GSS parameters to further minimise the constructed hierarchical state space. Since this process is very brief compared to the unfolding with CC, it could be worth to explore different parameter options here, although some could heavily impair runtimes (factorial complexity), so they're not pushed into the "factorial danger zone" at all by default.